### DISCUSSION TITLE
Open Sourcing checklist

### DISCUSSION THREAD

**Discussion by mdirolf** [crosshare-org/crosshare#248] [General]
I've invited a couple of folks as contributors to look over the codebase and get it ready for open sourcing. Let's use this discussion to coordinate

I'm not able to add fine-grained permissions for you so please refrain from pushing to master. If you have changes to make please use your own branch and make a PR when done - this will be more straightforward once the repo is public and folks can fork it rather than needing to be added as contributors.

I'd like to get it so that a new contributor can follow steps in the README in order to run the test suite and run the app locally. I'll try to add some basics to the README today but it would be awesome if you folks could try both of those things, ask questions where you get stuck, record any missing steps, etc.

Please use this discussion to note anything else you think we should do before making the source code public! I have done a rough audit for secrets and not found anything, but if you notice anything that you don't think ought to be public definitely speak up!

Thank you!

**Comment by mdirolf** [comment]
I've gone through and set up a local version of the project from scratch today, and tried to document as I went along in the README. Please give that a shot and feel free to edit/expand the readme (or ask any questions!) anywhere you think I'm missing some details.

It's a bit trickier to run locally than I'd imagined, as you have to create your own firebase project to connect to. Perhaps we could try to use the firebase emulator suite to bootstrap this for people running against a local version of firebase services instead of a live version?

**Reply by mdirolf** [reply]
Some more updates to the README - notably instructions for building the cluedb and building/uploading the wordlist to cloud storage. Hopefully these instructions are follow-able but it is a bit complicated for somebody trying to get started. I created this issue to try to remove that step: https://github.com/mdirolf/crosshare/issues/249

That being said - I was able to get the constructor working locally using the instructions that are now in the README - hopefully somebody else manages to confirm that as well!

**Reply by mdirolf** [reply]
Lastly, I added instructions for running the test suite. This should be very easy (famous last words) once you've managed to set up the local dev environment.

**Comment by jmviz** [comment]
I've followed the steps relatively without issue up to:

`$ GOOGLE_APPLICATION_CREDENTIALS=../serviceAccountKey.json ./scripts/bootstrapDatabase.ts
`

where I get an error: `Unexpected token ' in JSON at position 0`. Any idea what might be the issue there?

**Reply by mdirolf** [reply]
Glad to hear you made it so far! I wonder if your `serviceAccountKey.json` is the wrong file somehow or other. Can you take a look at the contents and make sure it looks like valid JSON? Mine looks something like:
```shell
$ cat serviceAccountKey.json 
{
  "type": "service_account",
  "project_id": "...",
  ...
}
```

**Reply by jmviz** [reply]
Yeah, that was my first though as well. The `serviceAccountKey.json` file appears to be valid json though and has the same structure as yours.

**Reply by mdirolf** [reply]
OK I think it might be that first line of `scripts/bootstrapDatabase.ts` where I pass in a json argument to ts-node-script. On my system I need the single quotes to get things to work, but it appears like maybe that is causing an issue on yours? Maybe try removing the single quote on either end of that json object and see what happens?

**Reply by mdirolf** [reply]
Hopefully this is something that would be simplified away by https://github.com/mdirolf/crosshare/issues/250

**Reply by jmviz** [reply]
That was it, thanks.

**Reply by mdirolf** [reply]
Glad we figured it out but now I gotta figure out why that's needed for me and not for you, heh! All of those files in the `scripts` dir have similar first lines so if you're running another it'll have the same issue until we sort that out, FYI

**Reply by jmviz** [reply]
Gotcha. I'm on macOS (zsh), fwiw.

**Reply by mdirolf** [reply]
Thanks! Oops - yeah - check out one of the older scripts from when I was on MacOS/zsh 😂 - https://github.com/mdirolf/crosshare/blob/ab749921b5600ae97fbe08f7a86ceea721945c74/app/scripts/addNotifications.ts#L1

**Reply by jmviz** [reply]
I was able to get crosshare running locally, including the constructor. When running `yarn test`, all tests but one pass. `Builder.test.tsx` fails. This is the beginning of the output for that test:

```
 FAIL  __tests__/Builder.test.tsx (20.067 s)
  ✕ moderate as daily mini (18050 ms)
  ○ skipped puzzle in progress should be cached in local storage
  ○ skipped publish as default
  ○ skipped change author name in publish dialogue should publish w/ new name
  ○ skipped publish custom / non-rectangular size

  ● moderate as daily mini

    TestingLibraryElementError: Unable to find an element with the text: Moderated!. This could be because the text is broken up by multiple elements. In this case, you can provide a function for your text matcher to make your matcher more flexible.
```

**Reply by mdirolf** [reply]
That is awesome news - thanks @jmviz!

I'm so ashamed to admit it but that test has actually been broken for a little while. The functionality it's supposed to be testing is working as intended and I could not figure out why it wasn't passing, so I decided to come back to it. It's probably time to actually do so!

**Reply by mdirolf** [reply]
Other than the issue you mentioned w/ `bootstrapDatabase.ts` was there anything you thought could be made clearer in the instructions or any other areas for improvement? I do hope the whole process can be streamlined with #249, #250, and #251

**Reply by jmviz** [reply]
I did have to unquote the json args in the other scripts as you mentioned. So finding a way to make those scripts cross-platform would be nice. Also, before running `npx firebase use --add`, I had to `export PATH="$(yarn bin):$PATH"` to find firebase and then `firebase login`. When running `npx firebase deploy --only firestore`, it took about 20 minutes for firebase to build everything; before that was done, doing `yarn dev` and opening `http://localhost:3000/` resulted in errors.

Those issues you link I think would definitely make it easier / more straightforward. For the wordlist in particular it seems to me it would make sense for it to already be compiled and ready to go, perhaps as part of the repo itself, or accessed from crosshare.org as #249 suggests.

**Reply by mdirolf** [reply]
Oo good to know about needing to add firebase to the path and login! Also you're right I should make a note about the index build time - I'm surprised it took so long but it definitely needs to finish.

Agreed that the wordlist feels like the lowest hanging fruit to improve. I am curious to spend some time with the emulators too and see if that could make things substantially simpler.

**Reply by mdirolf** [reply]
I believe this commit should fix the scripts issue - moved the tsconfig into a separate file in the scripts directory and removed those arguments altogether: https://github.com/mdirolf/crosshare/commit/afdccbbbe9a0021165840b8feb70a165e7ff848f

**Comment by TechnoCrepe** [comment]
I'm running on Windows, and I too roughly got to the same spot as @jmviz with running `GOOGLE_APPLICATION_CREDENTIALS=<stuff>`

At a glance, I think it would be possible to get it running locally on Windows, but it may require some alternatives (e.g. bootstrapDatabase.ts is currently treated like a bash script).

Before I forget, 2 things I did do while setting up was ...

- Install [nvm for windows](https://github.com/coreybutler/nvm-windows) and install node 12 via that. 
- I also needed to run `npx firebase login` prior to `npx firebase use --add`

It's currently 1 A.M. and I'm starting to re-read sentences a few times over, so I thought I'd jot this down and continue getting it to work tomorrow.

**Reply by mdirolf** [reply]
Awesome, thanks Sean! Yeah I imagine Windows will be a challenge. Unfortunately it's not one that I think I'll be able to lead the charge on - I just completely lack experience there - but I'd be happy to try to help however I can! I wonder if it'd be easier to run `bootstrapDatabase.ts` using ts-node directly rather than relying on the shebang at the top of the file. Something like `npx ts-node-script scripts/bootstrapDatabase.ts`. You'll need to set that `GOOGLE_APPLICATION_CREDENTIALS` environment var still - not sure if there's a different way of doing that on windows?

I'm going to add that `npx firebase login` to the README - @jmviz had the same issue - I must've just already been logged in when I tried to start a fresh environment.

Please keep the notes coming! It's very helpful! I do continue to wonder if the best way forward will be a docker image for development that can run cross platform.

**Reply by jmviz** [reply]
I was able to get it working on my Windows machine using [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/about), which I would recommend if it looks like a pure Windows setup will be a headache (I didn't try).

**Reply by mdirolf** [reply]
Awesome thanks @jmviz - I'm going to add that to the README

**Reply by TechnoCrepe** [reply]
Yeah, I debated WSL at first, then tried the pure-Windows route for a brief moment, before realizing it would take way too much effort 😛
But I was able to get WSL running the other night and got the local server going too

@mdirolf - I actually got a handful of test failures....

Something like...
```
Snapshot Summary
 › 3 snapshots failed from 2 test suites. Inspect your code changes or press `u` to update them.

Test Suites: 3 failed, 21 passed, 24 total
Tests:       4 failed, 197 passed, 201 total
Snapshots:   3 failed, 157 passed, 160 total
Time:        57.196 s
Ran all test suites.
```

I don't suppose both of you are on the East Coast? 
If so there may be a bug with the snapshots regarding timezones, as I am on the West Coast, for example....

![image](https://user-images.githubusercontent.com/82124832/114334553-a164ae00-9aff-11eb-958f-d4d38d17ebd1.png)

**Reply by TechnoCrepe** [reply]
It seems there is an open issue on date-fns/date-fns#2151, which is what is used to generate these strings. 
There are suggestions around (like this one https://github.com/date-fns/date-fns/discussions/2366#discussioncomment-532000) that suggest to just use Date's `.toISOString()` instead.

I'll try and fix all the snapshots that are have this since it's affecting me and send out a PR.

**Reply by mdirolf** [reply]
Great thanks @TechnoCrepe - yup I'm on east coast and haven't run them anywhere else so that seems likely to be the issue!

Semi-relatedly - the tests are definitely the portion of the codebase I'm least satisfied with. They're not nearly exhaustive, they are a bit flaky, and they're slow to boot. So any improvements there are greatly appreciated!

**Comment by mdirolf** [comment]
There are a couple of open issues that have come up following this discussion (#249, #250, #251), but since several of you have gotten things running and there haven't been any show-stopping discoveries about the code base, I'm leaning towards flipping the switch on opening the codebase soon. Anybody have any reasons we should wait?

**Reply by jmviz** [reply]
I say go for it :), more eyes will result in more improvements more quickly!

One small thing I noticed is there is a ToS page, bits of which could be updated to reflect the new open-source license. But it doesn't look like that page is accessible from any other page at the moment, so it is maybe a nonissue.

**Reply by TechnoCrepe** [reply]
Yeah I think it'll be fine. I think the "good first issue" tag is also a good idea to help folks get started with some direction.

**Reply by mdirolf** [reply]
OK thanks guys! I have flipped the switch so the repo should be public now 🥳. As such I've removed you folks as collaborators so that we can try what I think ought to be the more "normal" workflow - i.e. fork/PR for changes rather than pushing to this repo directly. I think I have migrated any action items from this discussion into normal issues But feel free to create any more if there's something I missed. Thanks again!
